src/btree/bt_page.c

```diff
@@ -120,249 +202,418 @@ __page_build_ref(WT_SESSION_IMPL *session, WT_REF *parent_ref, WT_CELL_UNPACK_AD
-    uint32_t i, j;
+    size_t pindex_size;
+    size_t refs_entries, incr;
+    uint32_t i;
 
-    unpacked_deltas = unpacked_deltas_merged = NULL;
     refs = NULL;
-    pindex = NULL;
-    unpacked_deltas_merged_size = refs_entries = incr = 0;
+    refs_entries = 0;
+    incr = 0;
 
-    /*
-     * !!!
-     * Unpack all delta images into a 2D array where entry is WT_CELL_UNPACK_DELTA_INT:
-     *     unpacked_deltas[0] -> [ entry_0_0, entry_0_1, entry_0_2, ... ]
-     *     unpacked_deltas[1] -> [ entry_1_0, entry_1_1, entry_1_2, ... ]
-     *     unpacked_deltas[2] -> [ entry_2_0, entry_2_1, entry_2_2, ... ]
-     *                           ...
-     *     unpacked_deltas[N] -> [ entry_N_0, entry_N_1, entry_N_2, ... ]
-     */
-    WT_RET(__wt_calloc_def(session, delta_size, &delta_size_each));
-    WT_ERR(__wt_calloc_def(session, delta_size, &unpacked_deltas));
-    for (i = 0, j = 0; i < (uint32_t)delta_size; ++i, j = 0) {
-        header = (WT_PAGE_HEADER *)deltas[i].data;
-        WT_ASSERT(session, header->u.entries != 0);
-        WT_ASSERT(session, header->u.entries % 2 == 0);
-
-        delta_size_each[i] = (size_t)header->u.entries / 2; /* Each entry has a key and a value. */
-        WT_ERR(__wt_calloc_def(session, header->u.entries, &unpacked_deltas[i]));
-        WT_CELL_FOREACH_DELTA_INT(session, ref->page->dsk, header, unpacked_deltas[i][j])
-        {
-            j++;
-        }
-        WT_CELL_FOREACH_END;
-    }
+    /* Merge deltas directly with the base image to build refs in a single pass. */
+    WT_RET(__page_merge_deltas_with_base_image_old(
+      session, ref, deltas, delta_size, &refs, &refs_entries, &incr));
 
-    WT_ERR(__page_merge_internal_deltas(session, unpacked_deltas, 0, (uint32_t)delta_size - 1,
-      delta_size_each, &unpacked_deltas_merged, &unpacked_deltas_merged_size));
-
-    WT_ERR(__page_merge_internal_delta_with_base_image(session, ref, unpacked_deltas_merged,
-      unpacked_deltas_merged_size, &refs, &refs_entries, &incr));
-
-    /*
-     * Constructs a new `p-index` using the merged `refs` list and allocate refs to the new
-     * `p-index`.
-     */
+    /* Allocate a new page index and assign refs to it. */
     pindex_size = sizeof(WT_PAGE_INDEX) + refs_entries * sizeof(WT_REF *);
     WT_ERR(__wt_calloc(session, 1, pindex_size, &pindex));
     incr += pindex_size;
+
     pindex->index = (WT_REF **)(pindex + 1);
     pindex->entries = (uint32_t)refs_entries;
+
     for (i = 0; i < pindex->entries; ++i) {
         refs[i]->pindex_hint = i;
         pindex->index[i] = refs[i];
     }
 
-    /* Initialize the reconstructed `p-index` into the internal page */
+    /* Install the reconstructed page index into the internal page. */
     WT_INTL_INDEX_SET(ref->page, pindex);
     __wt_cache_page_inmem_incr(session, ref->page, incr, false);
 
+    /* Cleanup the refs array (the actual WT_REFs are now owned by the page index) */
```

Comment:
Why we free the refs? Will this free the refs in the page?

Reply:
This is the current implementation while reconstructing internal deltas, this will go away with the new implementation.